src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs

```diff
@@ -10,14 +10,36 @@ namespace Microsoft.ML.Runtime.Internal.CpuMath
     public static partial class CpuMathUtils
     {
         // The count of bytes in Vector128<T>, corresponding to _cbAlign in AlignedArray
-        public const int Vector128Alignment = 16;
+        private const int Vector128Alignment = 16;
+
+        // The count of bytes in Vector256<T>, corresponding to _cbAlign in AlignedArray
+        private const int Vector256Alignment = 32;
+
+        public static int GetVectorAlignment()
+        {
+            // Assumes SSE support on machines that run ML.NET.
+            return Avx.IsSupported ? Vector256Alignment : Vector128Alignment;
```

Comment:
We shouldn't assume SSE support, this may break for ARM32/ARM64 when/if they come online.

Reply:
ML.NET currently assumes SSE3 support for the active existing intrinsics in `SseUtils` to function properly.  The original alignment value was a constant value 16.  When ARM32/ARM64 is adopted in the future, IMO those PRs should address this assumption, but not sure...

Reply:
It would be better to, IMO, assert that this method is only called if `Sse.IsSupported` is true then (as the HWIntrinsic code paths should not be getting called otherwise).

It is also undesirable to assume SSE3 support, as there are platforms/OS's where SSE3 support might not be available (I don't think I saw any code that assumed this condition, for the HWIntrinsics, however).

Reply:
I think adding `Contracts.Assert(Sse.IsSupported);` is good, but then (1) software fallbacks would fail at the assert even though they don't care about alignment, (2) external calls to `GetVectorAlignment` particularly in `RffTransform.cs` would fail before calling any software fallbacks, and (3) I'll need to add the same assert to `.netstandard.cs` (i.e. the original code path), which could be fine since those machines without SSE support would have failed to call SSE intrinsics anyway.

I agree with adding `Contracts.Assert(Sse.IsSupported)` somewhere, but also want to avoid the issues in (1) and (2), so I'd love to know about how I should proceed along this direction.

I agree that there may be some systems without SSE3 support.  It seems to me that the `Sse3.HorizontalAdd` intrinsic, which is frequently called for getting the `VectorSum`, is only available in SSE3 but unfortunately not in SSE, but I may be mistaken...

Reply:
To handle the software fallbacks (1), you should probably be returning the natural alignment of `float` (which is generally 4) or just not caring about alignment for that case in general (since the JIT will handle it for primitive types).

Reply:
That is to say, there is no point in copying the `AlignedArray` to be 16-byte aligned, when the software fallback will only be accessing elements that require 4-byte alignment.

And long term, we probably shouldn't even have AlignedArray, and should be properly handling the head/tail of the arrays (either via double-compute or scalar access) to ensure the rest of the data is aligned.

Reply:
Awesome! Will do this - thanks!

Note: In some other files like `CpuAligenedMathUtils` (yes it is spelled this way currently), they are checking if `a.cbAlign == CpuMathUtils.GetVectorAlignment`.  I will change it to `(a.cbAlign % CpuMathUtils.GetVectorAlignment) == 0` in my next wave of commits.

Reply:
I can't do the same checks `Sse/Avx.IsSupported` in the `GetVectorAlignment()` in `.netstandard.cs`, so I may assume SSE support there and always return `Vector128Alignment` since it was the original behavior.

Reply:
In all honesty, I plan on removing `AlignedArray` all together at some point.  So I wouldn't spend too much time polishing this, as it will go away.